controllers/pxc/perconaxtradbcluster_controller.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	o := &pxcv1.PerconaXtraDBCluster{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, o)
+	if err != nil {
+		return rr, err
+	}
+
+	if _, ok := o.Annotations["issue-vault-token"]; !ok {
+		return rr, nil
+	}
+
+	log.Info("found annotation")
+
+	newSecretObj := corev1.Secret{}
+	err = r.Client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: o.Namespace,
+			Name:      o.Spec.VaultSecretName,
+		},
+		&newSecretObj,
+	)
+	if !apierrors.IsNotFound(err) {
+		return rr, err
+	}
+	if err == nil {
+		log.Info("issued secret was found, delete annotaton")
+		err = r.deleteAnnotation(o)
+		return rr, err
+	}
+
+	rootSecretName, err := rootSecretName()
+	if err != nil {
+		return rr, err
+	}
+
+	rootSecretObj := corev1.Secret{}
+	err = r.Client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: o.Namespace,
+			Name:      rootSecretName,
+		},
+		&rootSecretObj,
+	)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("root secret was not found in namespace")
+			return rr, nil
+		}
+		return rr, err
+	}
+
+	err = r.IssueVaultToken(rootSecretObj, o.Spec.VaultSecretName)
+	if err != nil {
+		return rr, err
+	}
+
+	log.Info("token was issued, delete annotation")
+
+	return rr, r.deleteAnnotation(o)
+}
+
+func (r *PerconaXtraDBClusterReconciler) deleteAnnotation(o *pxcv1.PerconaXtraDBCluster) error {
+	return r.Client.Patch(context.Background(), o, client.RawPatch(types.JSONPatchType, []byte("[{\"op\": \"remove\", \"path\": \"/metadata/annotations/issue-vault-token\"}]")))
+}
+
+func rootSecretName() (string, error) {
+	if s, ok := os.LookupEnv("VAULT_SECRET_NAME"); ok {
+		return s, nil
+	}
+	return "", errors.New("VAULT_SECRET_NAME env is not set")
+}
+
+func (r *PerconaXtraDBClusterReconciler) IssueVaultToken(rootVaultSercet corev1.Secret, newSecretName string) error {
+	data := string(rootVaultSercet.Data["keyring_vault.conf"])
+	fields := strings.Split(data, "\n")
+	conf := make(map[string]string)
+	for _, f := range fields {
+		kv := strings.Split(f, "=")
+		conf[strings.TrimSpace(kv[0])] = strings.TrimSpace(kv[1])
+	}
+
+	newData := make(map[string][]byte)
+
+	tr := &http.Transport{}
+	if ca, ok := rootVaultSercet.Data["ca.cert"]; ok {
+		newData["ca.cert"] = ca
+
+		certPool, err := x509.SystemCertPool()
+		if err != nil {
+			return fmt.Errorf("failed to get system cert pool: %v", err)
+		}
+
+		ok := certPool.AppendCertsFromPEM(ca)
+		if !ok {
+			return fmt.Errorf("failed to append cert")
+		}
+
+		tr.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+			RootCAs:            certPool,
+		}
+	}
+
+	httpClient := &http.Client{
+		Timeout:   10 * time.Second,
+		Transport: tr,
+	}
+	cli, err := api.NewClient(&api.Config{
+		HttpClient: httpClient,
+		Address:    conf["vault_url"],
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create vault client: %v", err)
+	}
+
+	cli.SetToken(conf["token"])
+	path := fmt.Sprintf("%s/%s", conf["secret_mount_point"], rootVaultSercet.Namespace)
```

Comment:
it is possible to have multiple clusters in one namespace
```suggestion
	path := fmt.Sprintf("%s/%s/%s", conf["secret_mount_point"], newSecret.Namespace, newSecret.Name)
```

Reply:
https://github.com/Percona-Lab/k8s-vault-issuer/pull/1/commits/10393bf5f7de93dd13e69323c044f9126f2cdb7d